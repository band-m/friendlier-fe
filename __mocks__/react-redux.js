jest.mock('react-redux', () => {
  return {
    __esModule: true,
    useDispatch: jest.fn(() => () => {}),
    useSelector: jest.fn((selector) => selector({ contacts: { contactList: [] }, auth: { user: '' } }))
  };
});
