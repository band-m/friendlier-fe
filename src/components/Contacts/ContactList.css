@import url("https://fonts.googleapis.com/css?family=DM+Serif+Display|Hepta+Slab:300,400,600,900|Lato:300,400,700|Raleway:200,300,400,500,700,900|Rozha+One&display=swap");

.ContactList{
  display: flex;
  flex-direction: column;
  width: 25vw;
  max-height: 94vh;
  overflow: scroll;
  padding: 0;
  display: flex;
  align-items: center;
  background: #f7b82e;
  justify-content: flex-start;
  font-family: Raleway;
  border-right: 2px solid #0e4375;

  h2 {
    width: 100%;
    margin: 15px;
    font-size: 25px;
    text-align: center;
    color: #0e4375;
    border-right: 2px solid #0e4375;
  }

  li {
    height: 40px;
    width: 20vw;
    border-radius: 5px;
    border: 1px solid #0e4375;
    background: linear-gradient(#eeeeee, #8fdbf6);
    display: flex;
    align-items: center;
    justify-content: center;
    
    span {
      color: #0e4375;
      font-size: 1em;
    }
  }
  
  a{
    text-decoration: none;
    margin: 0.75rem auto;

    button{
      font-size: 2rem;
    }
  }
}

.sortedGreen {
  background-color: #64cf73;
}