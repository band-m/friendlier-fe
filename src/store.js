import { createStore, compose, applyMiddleware } from 'redux';
import promise from 'redux-promise-middleware';
import reducer from './redux/reducers/main-reducer';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default createStore(
  reducer,
  composeEnhancers(
    applyMiddleware(promise)
  )
);
