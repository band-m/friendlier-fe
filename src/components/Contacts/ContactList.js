import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { selectContactsList } from '../../data/selectors/contacts-selectors';
import { fetchContacts } from '../../data/actions/contacts-actions';
import isFuture from 'date-fns/isFuture';
import isPast from 'date-fns/isPast';
import differenceInCalendarDays from 'date-fns/differenceInCalendarDays';
import { selectUser, selectLoggedOut } from '../../data/selectors/auth-selector';
import styles from './ContactList.css';
import { SET_CURRENT_ZONE_RATIO } from '../../data/action-types/action-types';
import { myAction } from '../../data/actions/contact-detail-actions';
import { Link } from 'react-router-dom';

export default function ContactList() {
  const dispatch = useDispatch();
  const user = useSelector(selectUser);
  const loggedOut = useSelector(selectLoggedOut);

  useEffect(() => {
    if(user) {
      dispatch(fetchContacts(user._id));
    }
  }, [user]);

  useEffect(() => {
    if(loggedOut) {
      history.push('/');
    }
  }, [loggedOut]);

  const contacts = useSelector(selectContactsList);

  const greenZoneContacts = contacts.filter(contact => isFuture(new Date(contact.yellowZoneStartDate)));
  const yellowZoneContacts = contacts.filter(contact => isFuture(new Date(contact.redZoneStartDate)) && isPast(new Date(contact.yellowZoneStartDate)));
  const redZoneContacts = contacts.filter(contact => isFuture(new Date(contact.deadlineDate)) && isPast(new Date(contact.redZoneStartDate)));
  const overdueContacts = contacts.filter(contact => isPast(new Date(contact.redZoneStartDate)));

  useEffect(() => {
    greenZoneContacts.map(contact => {
      const daysIntoZone = contact.lastContactedDate ?
        differenceInCalendarDays(contact.lastContactedDate, Date.now()) :
        differenceInCalendarDays(contact.createdOn, Date.now());
      const ratio = daysIntoZone / contact.totalGreenZoneDays;
      dispatch(myAction(SET_CURRENT_ZONE_RATIO, ratio));
    });

    yellowZoneContacts.map(contact => {
      const daysIntoZone = contact.lastContactedDate ?
        differenceInCalendarDays(contact.yellowZoneStartDate, Date.now()) :
        differenceInCalendarDays(contact.yellowZoneStartDate, Date.now());
      const ratio = daysIntoZone / contact.totalYellowZoneDays;
      dispatch(myAction(SET_CURRENT_ZONE_RATIO, ratio));
    });

    redZoneContacts.map(contact => {
      const daysIntoZone = contact.lastContactedDate ?
        differenceInCalendarDays(contact.redZoneStartDate, Date.now()) :
        differenceInCalendarDays(contact.redZoneStartDate, Date.now());
      const ratio = daysIntoZone / contact.totalRedZoneDays;
      dispatch(myAction(SET_CURRENT_ZONE_RATIO, ratio));
    });
  }, []);

  const sortedGreen = greenZoneContacts.sort((a, b) => {
    a.zoneRatio - b.zoneRatio;
  });

  const sortedYellow = yellowZoneContacts.sort((a, b) => {
    a.zoneRatio - b.zoneRatio;
  });

  const sortedRed = redZoneContacts.sort((a, b) => {
    a.zoneRatio - b.zoneRatio;
  });

  const makeListItems = zone => {
    if(zone.length){
      return zone.map(contact => {
        return (
          <Link key={contact._id} to={`/contacts/${contact._id}`}>
            <li className={styles.commStatus}>
              <span>{contact.firstName} {contact.lastName}</span>
              {/* <span>{statusIcon}</span> */}
            </li>
          </Link>
        );
      });
    }
  };

  let greenZoners = makeListItems(sortedGreen);
  let redZoners = makeListItems(sortedYellow);
  let yellowZoners = makeListItems(sortedRed);
  let overdueZoners = makeListItems(overdueContacts);

  // let contactList;
  // if(contacts.length) {
  //   contactList = contacts.map(contact => {
  //     return (
  //       <Link key={contact._id} to={`/contacts/${contact._id}`}>
  //         <li className={styles.commStatus}>
  //           <span>{contact.firstName} {contact.lastName}</span>
  //           {/* <span>{statusIcon}</span> */}
  //         </li>
  //       </Link>);
  //   });
  // }

  return (
    <section className={styles.ContactList}>
      {contacts && <ul>
        {overdueZoners}
        {redZoners}
        {yellowZoners}
        {greenZoners}
      </ul>}
      {!contacts && <p>Add your first contact to get started!</p>}
      <Link to='/add'>
        <button title="Add New Contact">+</button>
      </Link>
    </section>
  );
}
