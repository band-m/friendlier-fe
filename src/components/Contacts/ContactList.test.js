import React from 'react';
import { shallow } from 'enzyme';
import ContactList from './ContactList';
import { Provider } from 'react-redux';
import store from '../../store';

describe('ContactList component', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<Provider store={store}><ContactList /></Provider>);
    expect(wrapper).toMatchSnapshot();
  });
});
