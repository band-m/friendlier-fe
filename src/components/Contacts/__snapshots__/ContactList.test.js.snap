// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList component matches snapshot 1`] = `
<Fragment>
  <ul />
  <Component
    to="contacts/add"
  >
    <img
      src="PLUS_ICON"
    />
  </Component>
</Fragment>
`;
